File: python/inheritance/library_management.py
"""
Problem Statement:
Design a Python program to manage a library system using inheritance.
Create a base class that stores common book details such as book ID and title.
Derive a class that adds author information, and further derive another class
that handles borrowing details. The program should accept details for multiple
books from the user and display complete borrowing information using multilevel
inheritance.
"""

class Book:
    def __init__(self, book_id, title):
        self.book_id = book_id
        self.title = title


class AuthorBook(Book):
    def __init__(self, book_id, title, author):
        super().__init__(book_id, title)
        self.author = author


class BorrowedBook(AuthorBook):
    def __init__(self, book_id, title, author, borrower_name, days):
        super().__init__(book_id, title, author)
        self.borrower_name = borrower_name
        self.days = days

    def calculate_fine(self):
        if self.days > 14:
            return (self.days - 14) * 5
        return 0


books = []
n = int(input("Enter number of borrowed books: "))

for i in range(n):
    print(f"\nBook {i + 1}")
    book_id = int(input("Enter book ID: "))
    title = input("Enter book title: ")
    author = input("Enter author name: ")
    borrower = input("Enter borrower name: ")
    days = int(input("Enter number of days borrowed: "))

    books.append(BorrowedBook(book_id, title, author, borrower, days))


print("\n--- Borrowed Book Details ---")
for book in books:
    print("\nBook ID:", book.book_id)
    print("Title:", book.title)
    print("Author:", book.author)
    print("Borrower:", book.borrower_name)
    print("Fine Amount:", book.calculate_fine())


"""
Sample Output:

Enter number of borrowed books: 2

Book 1
Enter book ID: 201
Enter book title: Clean Code
Enter author name: Robert Martin
Enter borrower name: Arun
Enter number of days borrowed: 10

Book 2
Enter book ID: 202
Enter book title: Design Patterns
Enter author name: Erich Gamma
Enter borrower name: Meera
Enter number of days borrowed: 18

--- Borrowed Book Details ---

Book ID: 201
Title: Clean Code
Author: Robert Martin
Borrower: Arun
Fine Amount: 0

Book ID: 202
Title: Design Patterns
Author: Erich Gamma
Borrower: Meera
Fine Amount: 20
"""

File: java/inheritance/LibraryManagement.java
/*
Problem Statement:
Design a Java program to manage a library system using inheritance.
Create a base class that stores common book details such as book ID and title.
Derive a class that adds author information, and further derive another class
that manages book issue details. The program should accept details for multiple
issued books from the user and display complete book and issue information
using multilevel inheritance.
*/

import java.util.Scanner;

class Book {
    protected int bookId;
    protected String title;

    public Book(int bookId, String title) {
        this.bookId = bookId;
        this.title = title;
    }
}

class AuthorBook extends Book {
    protected String author;

    public AuthorBook(int bookId, String title, String author) {
        super(bookId, title);
        this.author = author;
    }
}

class IssuedBook extends AuthorBook {
    private String issuedTo;
    private int days;

    public IssuedBook(int bookId, String title, String author, String issuedTo, int days) {
        super(bookId, title, author);
        this.issuedTo = issuedTo;
        this.days = days;
    }

    public int calculateFine() {
        if (days > 14) {
            return (days - 14) * 5;
        }
        return 0;
    }

    public void displayDetails() {
        System.out.println("Book ID: " + bookId);
        System.out.println("Title: " + title);
        System.out.println("Author: " + author);
        System.out.println("Issued To: " + issuedTo);
        System.out.println("Fine Amount: " + calculateFine());
    }
}

public class LibraryManagement {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of issued books: ");
        int n = sc.nextInt();

        IssuedBook[] books = new IssuedBook[n];

        for (int i = 0; i < n; i++) {
            System.out.println("\nBook " + (i + 1));
            System.out.print("Enter book ID: ");
            int id = sc.nextInt();
            sc.nextLine();

            System.out.print("Enter book title: ");
            String title = sc.nextLine();

            System.out.print("Enter author name: ");
            String author = sc.nextLine();

            System.out.print("Enter issued to: ");
            String issuedTo = sc.nextLine();

            System.out.print("Enter number of days issued: ");
            int days = sc.nextInt();

            books[i] = new IssuedBook(id, title, author, issuedTo, days);
        }

        System.out.println("\n--- Issued Book Details ---");
        for (IssuedBook book : books) {
            System.out.println();
            book.displayDetails();
        }

        sc.close();
    }
}

/*
Sample Output:

Enter number of issued books: 2

Book 1
Enter book ID: 101
Enter book title: Clean Code
Enter author name: Robert Martin
Enter issued to: Arun
Enter number of days issued: 10

Book 2
Enter book ID: 102
Enter book title: Design Patterns
Enter author name: Erich Gamma
Enter issued to: Meera
Enter number of days issued: 20

--- Issued Book Details ---

Book ID: 101
Title: Clean Code
Author: Robert Martin
Issued To: Arun
Fine Amount: 0

Book ID: 102
Title: Design Patterns
Author: Erich Gamma
Issued To: Meera
Fine Amount: 30
*/
